perfkitbenchmarker/linux_benchmarks/cloudsuite_in_memory_analytics_benchmark.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2016 PerfKitBenchmarker Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Runs the in-memory analytics benchmark of Cloudsuite.
+
+More info: http://cloudsuite.ch/inmemoryanalytics/
+"""
+
+import re
+
+from perfkitbenchmarker import configs
+from perfkitbenchmarker import flags
+from perfkitbenchmarker import sample
+from perfkitbenchmarker import vm_util
+
+flags.DEFINE_string('cloudsuite_in_memory_analytics_dataset',
+                    '/data/ml-latest-small',
+                    'Dataset to use for training.')
+flags.DEFINE_string('cloudsuite_in_memory_analytics_ratings_file',
+                    '/data/myratings.csv',
+                    'Ratings file to give the recommendation for.')
+FLAGS = flags.FLAGS
+
+BENCHMARK_NAME = 'cloudsuite_in_memory_analytics'
+BENCHMARK_CONFIG = """
+cloudsuite_in_memory_analytics:
+  description: >
+    Run Cloudsuite in-memory analytics benchmark. Specify the number of worker
+    VMs with --num_vms.
+  vm_groups:
+    master:
+      vm_spec: *default_single_core
+      vm_count: 1
+    workers:
+      vm_spec: *default_single_core
+"""
+
+
+def GetConfig(user_config):
+  config = configs.LoadConfig(BENCHMARK_CONFIG, user_config, BENCHMARK_NAME)
+  if FLAGS['num_vms'].present:
+    config['vm_groups']['workers']['vm_count'] = FLAGS.num_vms
+  return config
+
+
+def _HasDocker(vm):
+  resp, _ = vm.RemoteCommand('command -v docker',
+                             ignore_failure=True,
+                             suppress_warning=True)
+  return bool(resp.rstrip())
+
+
+def Prepare(benchmark_spec):
+  """Install docker. Pull the required images from DockerHub. Create datasets.
+  Start Spark master and workers.
+
+  Args:
+    benchmark_spec: The benchmark specification. Contains all data that is
+        required to run the benchmark.
+  """
+  master = benchmark_spec.vm_groups['master'][0]
+  workers = benchmark_spec.vm_groups['workers']
+
+  def PrepareCommon(vm):
```

Comment:
In python, there's a pickle module that can serialize and deserialize objects. One of its weaknesses is that it can't handle nested functions. We currently only use pickle to serialize the BenchmarkSpec to preserve state between stages of benchmark execution: provision, prepare, run, cleanup, teardown. Since the BenchmarkSpec doesn't hold any references to these functions, it's not required, just a recommendation. I'd be fine with leaving them like this.